[Cooper] 미션5 : 보드에 위치 부여 및 점수 계산

pbg0205 opened the pull request:
안녕하세요!
2021 코드스쿼드 마스터즈 백엔드 과정의 Cooper 입니다.🚀
하단에 주요 변경 사항 및 고민했던 부분에 관해서 정리해보았습니다.
사소하게라도 부족한 부분 있으면 거침없이 피드백 부탁드립니다!🔥
리뷰해주시는데 시간 내주셔서 감사합니다!
잘부탁드립니다😃
​
### 주요 변경 사항 & 고민했던 부분
1. PieceGroup의 자료구조를 Map<Color,List<Piece>> -> List<Piece>로 변경.
- 이전 미션에서 관리하는 Map<Color, List<Piece>>을 List<Piece>로 변경한 이유는 실제 체스판의 좌표와 같이 관리하기 위함 입니다.
- 체스판은 8 x 8 형태로 구성되며, 총 64개의 좌표를 가지고 있습니다. 이를 List의 인덱스를 이용하면 실제 실제판에서 사용하는 방법과 같이 체스판을 관리할 수 있는 장점이 있다고 생각했습니다.
- 일차원 형태를 2차원 형태로 사용하기 위해 인덱스의 몫은 행(row), 인덱스의 나머지는 열(column)으로 계산하여 각 좌표를 관리하고 있습니다.
​
2. 위치 정보를 관리하는 Position 클래스 생성
- 기존 구현 방식은 입력받는 위치 문자열 데이터를 파싱해 행과 열의 정보를 primitive 형태로 관리하고자 했지만 값을 주고 받는 과정에서 메서드의 매개인자가 많아지고 위치라는 하나의 의미로 값을 주고 받는 면이 효율적이라고 생각해 별도의 Position 클래스를 구현했습니다.
- 현재 단계에서는 동등성(equals)에 관한 요구사항이 없어 별도로 equals와 hashCode 재정의를 진행하지 않았습니다. 추후 관련 요구사항 확인 시, 재정의 작업 진행하겠습니다.
​
3. PieceGroup 멤버로 PointCalculator 생성
- 기물의 점수를 계산하는 책임을 PointCalculator에 부여 했습니다.
- 기물에 대한 정보를 PieceGroup이 관리하고 있어 내부에 계산 기능을 추가하되, 계산 기능을 담당하는 클래스를
분리시키는 것이 적합다하고 생각했습니다. 클래스로 분리할 경우 계산 역할을 하는 코드를 별도로 관리할 수 있고 계산 기능에 대한 책임을 적용하는데 적합할 것 같은데 올바른 생각인지 궁금합니다.
​
4. 기물의 점수에 따른 정렬 기능 구현
- 체스판에 존재하는 기물을 기물의 할당된 점수에 따라 오름차순 정렬하는 기능을 구현했습니다. 구현 시, Piece에 Comparable을 구현하는 방법을 고민했지만 해당 방법을 사용할 경우, 한가지의 경우만 정렬이 가능하여 다른 정렬 방법을 적용하는데 어려움이 있을 것이라고 판단했습니다.
- 그래서 PieceGroup에서 전달한 색상에 따른 리스트를 Comparator.comparingDouble을 사용하였습니다. 각가의 리스트에 다른 정렬 방식(오름차순, 내림차순)을 적용해 정렬된 리스트를 반환하였습니다.​

pbg0205 committed:
style : getPawnList 메서드 명 변경

- 기존에 변경하지 못한 getPawnList를 getPieceList로 메서드 명을 변경.

- PieceGroup add 메서드 적용 가능 범위 재설정.

pbg0205 committed:
feat : Color 상수 NO_COLOR 추가

pbg0205 committed:
feat : Type Enum 상수 NO_PIECE 추가

- NO_PIECE의 멤버로 '.'을 추가해 getRepresenation에서 NO_COLOR가
입력될 때 NO_PIECE.representation을 반환하도록 구현했습니다.

pbg0205 committed:
test : representation 확인 테스트 코드 추가

- Type마다 해당하는 representation을 확인하기 위한 테스트 코드를 작성
하였습니다.

pbg0205 committed:
feat : getType과createBlank 메서드 추가

pbg0205 committed:
feat : isWhite & isBlack 메서드 추가

pbg0205 committed:
test : create_piece & create_blank 메서드 추가

- 기존의 representation과 Color로 확인했던 테스트 코드를 Type과 Color로
적합한 기물의 생성 여부를 확인하도록 변경하였습니다.

- 빈칸(blank) 생성 여부를 확인하는 create_blank 메서드를 추가
하였습니다.

pbg0205 committed:
refactor : 팩토리 메서드 중복 제거 메서드 추가

- 기존의 팩토리 메서드는 생성자를 통해 인스턴스를 전달하는 방식을
createWhite & createBlack 메서드를 선언해서 생성자 코드의 중복을 줄이
도록 작성하였습니다.

pbg0205 committed:
feat : PieceGroup 자료구조 변경

- 기존 Map<Color, List<Piece>> 자료 구조를
List<Piece> 형태로 변경하였습니다.

pbg0205 committed:
refactor : 사용하지 않은 메서드 제거

- 미션5부터 add 메서드를 사용하지 않는 add & findPawn
메서드를 제거하였습니다.

- 자료구조를 변경하면서 인덱스의 목을 행으로 취급하고
각 행의 상태를 출력 하였습니다.

pbg0205 committed:
feat : Color & Type 기물 갯수  메서드 추가

- 원하는 색상과 기물의 종류를 입력해 해당
기물의 갯수를 반환하는 메서드를 추가하였습니다.

pbg0205 committed:
test : getPieceCount 테스트 메서드 추가

- 색생 및 기물 종류에 따른 기물의 갯수를 확인하는 테스트 코드를
작성하였습니다.

pbg0205 committed:
fix : equals 재정의 코드 수정

- (o instanceof Piece) -> !(o instanceof Piece) 변경

pbg0205 committed:
feat : board 클래스 findPiece 메서드 추가

- getXPos를 통해 체스판의 열의 위치를 반환하고
getYPos를 통해 체스판이 행의 위치를 반환하도록 구현했습니다.

- 애매해진 Board의 역할을 체스판의 좌표와 연관된 연산을
반환하는 채임을 부여하는 방식으로  계획하고 있습니다.

pbg0205 committed:
feat : PieceGroup -> findPiece 메서드 추가

- board클래스에서 반환한 행과 열의 값을 이용해
해당 위치에 있는 Piece를 전달하는 방식으로 구현했습니다.

pbg0205 committed:
test : findPiece 테스트 코드 추가

- 입력한 position에 정확한 기물이 있는지 확인하는 테스트
코드를 작성했습니다.

pbg0205 committed:
feat : 원하는 위치 기물 이동 메서드 추가

- move method를 확인하기 위해 initializeEmpty
메서드를 선언하여 체스판을 초기화 하였습니다.

- move 메서드를 선언하여 해당 기물을 원하는
위치로 이동할 수 있도록 구현 하였습니다.

pbg0205 committed:
test : board 이동 메서드 테스트 코드 추가

- 정확한 위치로 기물의 이동 여부를 확인하는
테스트 코드를 추가하였습니다.

pbg0205 committed:
feat : 위치를 전달하는 Position 클래스 생성

- 기존에 위치를 관리하던 Board 클래스의 책임은 x, y좌표만으로
관리하는 것이 한계가 있었습니다. 그래서 위치만을 관리할 수 있는
Position 클래스를 선언하였습니다.

pbg0205 committed:
feat : addPiece 및 move 메서드 구현

- 임의의 기물을 추가하는 메서드인 addPiece를 구현하였습니다.

- 이전 위치와 옮겨지는 위치를 통해 기물의 이동 기능을 하는
move 메서드를 선언 하였습니다. 해당 메서드를 구현하는 과정에서
x, y만으로 자료를 관리하는 것보다 Position으로 위치를 관리하는
것이 역할을 명확히하고 관리에 이점이 있어 Position을 선언한
이유이기도 합니다.

pbg0205 committed:
test : move & addPiece 메서드 테스트 코드 추가

pbg0205 committed:
feat : Type Enum 내 defaultPoint 추가

- 체스 프로그램에 점수를 계산하기 위해
멤버로 defaultPoint를 추가 하였습니다.

pbg0205 committed:
feat : 체스판 계산기능 구현

- Piece를 관리하는PieceGroup에서 계산 기능이 구현되어야 한다고
생각했습니다. 그래서 내부의 멤버변수로 PointCalculator로 선언하였
고 PointCalculator 내부에서 점수 계산의 메서드를 구현했습니다.

- PieceCalculator에서 내부에서 우선적으로 열을 탐색 후, 다음
행을 탐색하는 로직으로 메서드를 구현했습니다.

- 같은 열에 폰이 2개 이상 존재하는 경우의 연산을 대비해서
pawnCount의 갯수를 기준으로 연산을 다르게 수행하도록 했습니다.

pbg0205 committed:
test : 계산 기능을 확인하는 테스트 코드 작성

pbg0205 committed:
feat : 점수에 따른 정렬된 리스트 메서드 추가

- 점수의 오름차순으로 Board클래스에서 리스트를 정렬해서
반환하는 메서드를 선언했습니다.

- 점수의 내림차순으로 리스트를 반환하는 메서드를 선언했습니다.

- Piece에 comparable을 구현하는 방법을 쓸지에 대해 고민했지만
다양한 정렬 방식을 구현하기 위해서는 sort 내부에 정렬 기준을
작성하는 것이 적합하다고 생각했습니다.

pbg0205 committed:
test : 점수의 오름차순, 내림차순 정렬 테스트

SANGCO commented on src/main/java/cooper/chess/piece/PieceGroup.java:
```diff
@@ -1,28 +1,34 @@
+    private static int BLACK_PIECE_SIZE;
 
-    private final Map<Color, List<Piece>> pieceListMap;
+    private final List<Piece> pieceList;
```
`PieceGroup` 이라는 클래스를 만들었으니 이 클래스를 좀 더 활용해 봅시다.
지금은 하나의 `PieceGroup`에 모든 `Piece`들을 관리하고 있는데
비슷한 속성을 가진 `Piece`들을 모아 하나의 `PieceGroup`으로 만들봅시다.
이런 식으로 여러개의 `PieceGroup`을 만들어 봅시다.
그러면 `Board` 클래스가 여러개의 `PieceGroup'을 들고있는 구조가 되겠죠?

pbg0205 replied:
리뷰 감사합니다! 이번에 수정을 하는 과정에 있어 자료 구조에 따라 변경사항이 많이 바뀔꺼 같습니다. 그래서 수정 진행하기 전 현재 구상하고 있는 접근 방식에 대해 먼저 말씀 드리고 수정하는 편이 맞는 것 같아 리뷰 남깁니다ㅎㅎ
 PieceGroup의 key-value를 <Position, Piece> 형태로 선언해서 모든 Position 관리하는 방법이 기물을 이동하는데  유용할 것으로 생각합니다 . 이를 기준으로 해서 PieceGroup에서 점수 연산의 책임을 부여하는 방식으로 구상하고 있습니다. 이 접근 방법은 어떻게 생각하실까요??

SANGCO replied:
이번에도 하나의 PieceGroup이 `Map<Position, Piece>` 형태로 모든 Piece들을 관리하는 건가요?

pbg0205 replied:
가장 중점적으로 기준을 선정하는 부분이 전체 출력에 초점을 두고 생각을 하는 것 같습니다ㅠ 1차 대안은 Map<Position, Piece>로 생각해보았고 두 번째 방안으로 가로를 기준으로 하는 List<PiceGroup(Rank)>의 형태로 방법으로 고려하고 있습니다!

SANGCO replied:
두 번째 방안으로 한번 가보시죠!
첫번째꺼는 아마 처음하고 비슷하게 결과물이 나올거 같아서요.
Rank라는 클래스에 역할을 한번 부여해 봅시다.

pbg0205 replied:
네! Rank 클래스로 수정해서 리뷰 요청 드리겠습니다! 늦은 시간에도 피드백 주셔서 감사합니다😁

SANGCO commented on src/main/java/cooper/chess/piece/PointCalculator.java:
```diff
@@ -0,0 +1,56 @@
+
+import java.util.List;
+
+public class PointCalculator {
```
계산하는 클래스가 따로 필요할까요?

데이터를 한군데 모아 놓고 `if문`을 사용해서 계산을 하는게 아니라
데이터를 가진 객체에게 각자가 자신이 가진 데이터를 달라고 하는거죠.
계산이 필요하면 계산을 해서 달라고 하는거고.

현재 각각의 'Piece` 객체들은 자신의 색깔, 타입, 점수를 알고 있습니다.
`Position` 클래스를 만드셨던데 각각의 `Piece`들이 `Position` 클래스를 가지고 있는 구조로 만들어서
`Piece`들이 자신의 위치를 알게끔 만드는건 어떨까요?
그러면 `Piece`는 자신의 색깔, 타입, 점수, 위치 정보까지 다 가지고 있게 됩니다.

> pawn의 기본 점수는 1점이다. 하지만 같은 세로줄에 같은 색의 폰이 있는 경우 1점이 아닌 0.5점을 준다.

이 부분은 어떻게 해결하면 좋을까요?
`Piece`는 자신의 색깔을 알고 있습니다.
`Piece`는 자신이 pawn인지 아닌지 알 수 있습니다.
`Piece`는 자신의 위치를 알고 있어서 어느 위치에 어느 색깔에 pawn이 오면 점수가 0.5점이 되는지 알수 있습니다.


위에 드렸던 피드백 처럼 어떤 특성들을 `PieceGroup`으로 만들지에 따라서 이 부분의 구현도 많이 달라지겠죠?
기물의 종류, 위치, 색깔, 가로, 세로 등등 어떤 기준으로 `Piece`를 묶을지 고민해 봅시다.
도출해낸 결과에 따라 `PieceGroup`이라는 이름도 바뀔 수 있겠죠?

SANGCO commented on src/main/java/cooper/chess/piece/PieceGroup.java:
```diff
@@ -1,28 +1,34 @@
 
 public class PieceGroup {
-    public static final int PIECE_LIST_MAX = 16;
+    private static int WHITE_PIECE_SIZE;
+    private static int BLACK_PIECE_SIZE;
```
이름은 대문자로 상수인데 `final`이 빠졌네요?
상수를 선언하면서 바로 초기화 합시다.

SANGCO commented on src/main/java/cooper/chess/piece/PieceGroup.java:
```diff
@@ -54,36 +66,59 @@ private void initBlackPiece() {
 
-    public List<Piece> getPawnList(Color color) {
-        return pieceListMap.get(color);
+    private void setWhiteSize() {
+        WHITE_PIECE_SIZE = 16;
+    }
+
+    private void setBlackSize() {
+        BLACK_PIECE_SIZE = 16;
+    }
```
필요없겠죠?

SANGCO commented on src/main/java/cooper/chess/piece/PointCalculator.java:
```diff
@@ -0,0 +1,56 @@
+import java.util.List;
+
+public class PointCalculator {
+    private final List<Piece> pieceList;
```
`PointCalculator` 클래스를 없애지 않는다고 해도
`PointCalculator` 클래스가 `pieceList`를 가지고 있을 필요가 있을까요?
파라미터로 그냥 넘겨주면 되지 않을까요?

SANGCO reviewed (changes requested):
Cooper님 구현 잘하셨네요. 고생하셨습니다!

리뷰를 남겨 드린 부분은 같이 고민해 봅시다.
댓글 언제나 환영입니다~

pbg0205 committed:
feat : Position 멤버변수 final 추가

pbg0205 committed:
feat : Rank 클래스 추가 및 Board 로직 변경

- 기존의 PieceGroup 클래스를 행의 기물 정보를 담당하는 Rank 클래스로
변경하였습니다.

- PointCalculator에 부여되었던 계산 역할을 Board로 변경하였습니다.

pbg0205 committed:
feat : Rank 클래스 불필요 메서드 제거

pbg0205 committed:
test : BoardTest 테스트 값 변경 및 메서드명 변경

- 잘못된 테스트 값을 7.0을 수정하였습니다.

- getPieceCount의 의미를 명확하게 하기 위해
getEachPieceCount로 메서드 명을 변경하였습니다.

pbg0205 committed:
refactor : Piece 클래스 getDefaultPoint() 생성

- 불필요한 메서드 호출을 줄이기 위해 getDefaultPoint() 생성

pbg0205 committed:
feat : Rank 점수 연관 메서드 추가

pbg0205 commented:
이전에 말씀해주신대로 Rank 클래스로 변경하여 코드 수정했습니다.
부족하거나 아쉬운 부분 있으시면 거침없는 피드백 부탁드립니다!
감사합니다!

SANGCO commented on src/main/java/cooper/chess/piece/Rank.java:
```diff
@@ -0,0 +1,66 @@
+
+    public Rank(List<Piece> pieceList) {
+        this.pieceList = pieceList;
+        Collections.unmodifiableList(pieceList);
```
반환값을 받아서 사용할 게 아니면 이 코드는 왜 필요한건가요~

SANGCO commented on src/main/java/cooper/chess/Board.java:
```diff
@@ -1,59 +1,193 @@
+        List<Piece> list = getListOf(color);
+
+        list.sort(Comparator.comparingDouble(piece
+                -> - piece.getDefaultPoint()));
```
여기 `-` 마이너스는 왜 필요한건가요?
`Comparator.comparingDouble(Piece::getDefaultPoint).reversed()`
요거 하시려고 하셨던게 아닌가 싶기도 하네요~

pbg0205 replied:
아 맞습니다! reversed가 있는걸 처음 알았네요ㅠ 수정하도록 하겠습니다!🔥🔥

honux77 replied:
오 상코 꼼꼼하네요. 상꼼~

SANGCO commented on src/main/java/cooper/chess/Board.java:
```diff
@@ -1,59 +1,193 @@
+    public double calculatePoint(Color color) {
+        double result = 0;
+        for (int column = 0; column < BOARD_SIZE; column++) {
+            result += havePawnsInColumn(column, color) ?
+                    addPointInColumn(column, color)
+                            - countPawnCount(column, color) * 0.5
+                    : addPointInColumn(column, color);
```
이거는 리뷰는 아니고 개인 의견입니다.
간단한건 삼항 연산자 사용하면 깔끔한데 이 경우는 제 시각에선 별로 안깔끔해 보이네요.

pbg0205 replied:
if문으로 변경해보니까 의미도 더 잘보이고 깔끔해지네요ㅎㅎ 이 부분 다음 미션에서 수정하도록 하겠습니다!
디테이한 부분까지 봐주셔서 감사합니다!

SANGCO reviewed (approved):
Cooper님 수정 너무 잘하셨네요.
고생하셨습니다~
몇가지 간단한건 다음 미션 진행 하시면서 살짝 수정하면 될거 같습니다~
